pkg/jx/cmd/step/syntax/step_syntax_canonical.go

```diff
@@ -0,0 +1,414 @@
+	opts.StepOptions
+
+	Pack              string
+	BuildPackURL      string
+	BuildPackRef      string
+	Context           string
+	CustomImage       string
+	DefaultImage      string
+	NoKaniko          bool
+	KanikoImage       string
+	ProjectID         string
+	DockerRegistry    string
+	DockerRegistryOrg string
+	SourceName        string
+	CustomEnvs        []string
+
+	PodTemplates map[string]*corev1.Pod
+
+	GitInfo         *gits.GitRepository
+	VersionResolver *opts.VersionResolver
+}
+
+// NewCmdStepSyntaxCanonical Creates a new Command object
+func NewCmdStepSyntaxCanonical(commonOpts *opts.CommonOptions) *cobra.Command {
+	options := &StepSyntaxCanonicalOptions{
+		StepOptions: opts.StepOptions{
+			CommonOptions: commonOpts,
+		},
+	}
+
+	cmd := &cobra.Command{
+		Use:   "canonical",
+		Short: "Outputs a canonical representation of the pipeline to be executed",
+		Long:  "Reads the appropriate jenkins-x.yml, depending on context, from the current directory, if one exists, and outputs a canonical representation of the pipelines",
+		Run: func(cmd *cobra.Command, args []string) {
+			options.Cmd = cmd
+			options.Args = args
+			err := options.Run()
+			helper.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().StringArrayVarP(&options.CustomEnvs, "env", "e", nil, "List of custom environment variables to be applied to resources that are created")
+
+	options.AddCommonFlags(cmd)
+	return cmd
+}
+
+// AddCommonFlags adds common CLI options
+func (o *StepSyntaxCanonicalOptions) AddCommonFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&o.Pack, "pack", "p", "", "The build pack name. If none is specified its discovered from the source code")
+	cmd.Flags().StringVarP(&o.BuildPackURL, "url", "u", "", "The URL for the build pack Git repository")
+	cmd.Flags().StringVarP(&o.BuildPackRef, "ref", "r", "", "The Git reference (branch,tag,sha) in the Git repository to use")
+	cmd.Flags().StringVarP(&o.Context, "context", "c", "", "The pipeline context if there are multiple separate pipelines for a given branch")
+	cmd.Flags().StringVarP(&o.ServiceAccount, "service-account", "", "tekton-bot", "The Kubernetes ServiceAccount to use to run the pipeline")
+	cmd.Flags().StringVarP(&o.SourceName, "source", "", "source", "The name of the source repository")
+	cmd.Flags().StringVarP(&o.CustomImage, "image", "", "", "Specify a custom image to use for the steps which overrides the image in the PodTemplates")
+	cmd.Flags().StringVarP(&o.DefaultImage, "default-image", "", syntax.DefaultContainerImage, "Specify the docker image to use if there is no image specified for a step and there's no Pod Template")
+	cmd.Flags().BoolVarP(&o.NoKaniko, "no-kaniko", "", false, "Disables using kaniko directly for building docker images")
+	cmd.Flags().StringVarP(&o.KanikoImage, "kaniko-image", "", syntax.KanikoDockerImage, "The docker image for Kaniko")
+	cmd.Flags().StringVarP(&o.ProjectID, "project-id", "", "", "The cloud project ID. If not specified we default to the install project")
```

Comment:
`PipelineConfig.modifyStep`.

Reply:
I see, also for this Kaniko stuff. Hmm, seems odd and GCLoud specific. Maybe Kaniko only works for GCLoud for now?

Reply:
On further thought, would as a parameter, not something like `--kaniko-cache-repo` make more sense? That's really what it is used for in the end. If you need you can specify this explicitly, otherwise we build the default by reading the project id from the config map.

Reply:
Let's try to figure out what the rationale behind all this stuff is and rework 'em as a whole rather than just rewording bandaids for now?